platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/files/DataFileWriter.java

```diff
@@ -211,20 +162,127 @@ public synchronized long storeDataItem(final Consumer<BufferedData> dataItemWrit
-        // the buffer will be closed below anyway
-        metadata.updateDataItemCount(writingHeaderPbjData, dataItemCount);
+        // note that updateDataItemCount() messes up with writing buffer state (position), but the buffer will be closed
+        // below anyway
+        metadata.updateDataItemCount(headerBuffer.bufferedData, dataItemCount);
         // release all the resources
-        MemoryUtils.closeMmapBuffer(writingHeaderMmap);
-        MemoryUtils.closeMmapBuffer(writingMmap);
+        headerBuffer.close();
+        dataBuffer.close();
+
+        file.setLength(totalFileSize);
+        file.close();
+        closed = true;
+    }
+
+    @Override
+    public long getOffHeapConsumption() {
+        if (closed) {
+            return 0;
+        }
+        // Return the size of the data buffer even if not fully used but claimed to be used.
+        // Buffer is released and not used after completing writing into the file, but we probably want to track
+        // possible memory usage
+        // for each data file writer instance created.
+        // During writing (and it is fast because sequential), we might not see benefit of reporting actual busser space
+        // taken.
+        return dataBuffer.bufferSize;
+    }
+
+    /**
+     * Helper class to encapsulate logic of writing data into the file using memory mapped buffers and moving buffer when needed.
+     */
+    private class MovingBuffer implements Closeable {
+
+        private final long bufferSize;
+        private long fileStartOffset = 0;
+        private MappedByteBuffer bufferFileMapping;
+        private BufferedData bufferedData;
+
+        public MovingBuffer(long fileStartOffset, long bufferSize) throws IOException {
+            this.bufferSize = bufferSize;
+            move(fileStartOffset);
+        }
 
-        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.READ, StandardOpenOption.WRITE)) {
-            channel.truncate(totalFileSize);
-            // after finishWriting(), mmapPositionInFile should be equal to the file size
-            mmapPositionInFile = totalFileSize;
+        /**
+         * Maps the writing byte buffer to the given position in the file.
+         * Previous mapped byte buffer, if not null, is released.
+         *
+         * @param fileStartOffset new mapped byte buffer position in the file, in bytes
+         * @throws IOException if I/O error(s) occurred
+         */
+        private void move(long fileStartOffset) throws IOException {
+            this.fileStartOffset = fileStartOffset;
+
+            // ensure we have disk space buffer can use
+            file.setLength(fileStartOffset + bufferSize);
+
+            final MappedByteBuffer newMapping = file.getChannel().map(MapMode.READ_WRITE, fileStartOffset, bufferSize);
+            if (newMapping == null) {
+                throw new IOException("Failed to map file channel to memory");
+            }
+
+            close();
```

Comment:
This is correct, but looks weird anyway. The object is not closed here, it's the previous mapped buffer that is closed. What about having a private helper method like "closeCurrentMappedBuffer()" and calling it here and in close()?